patterns/interpreter.md

```diff
@@ -0,0 +1,185 @@
+production rules. For example, to derive `()()` we apply the following rules:
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
+
+## Context Free Grammar for our problem
+
+Recall that our task is translate infix expressions into postfix ones. Let's
+define a context free grammar for a set of infix expressions over `0,...,9, +,-`,
+where:
+
+- terminal symbols: `0,...,9, +,-`
+- non-terminal symbols: `exp, term`
+- start symbol is `exp`
+- and the following are production rules
+
+```ignore
+exp -> exp + term
+exp -> exp - term
+exp -> term
+term -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+In fact this grammar should be further transformed depending on what we are going
+to do with the grammar. For example, we might need to remove left recursion.
+For more details please see [Compilers:Principles,Techniques, and Tools](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)
+(aka Dragon Book).
+
+## Solution
+
+We simply implement a recursive descent parser. The code panics when an expression
+is syntactically wrong (for exmple `2-34` or `2+5-` are wrong according to the
+grammar definition).
+
+```rust
+pub struct Interpreter<'a> {
+    it: std::str::Chars<'a>,
+}
+
+impl<'a> Interpreter<'a> {
+
+    pub fn new(infix: &'a String) -> Self {
+        Self { it: infix.chars() }
+    }
+
+    pub fn interpret(&mut self, out: &mut String) {
+        self.exp(out);
+    }
+
+    fn next_char(&mut self) -> Option<char> {
+        self.it.next()
+    }
+
+    fn exp(&mut self, out: &mut String) {
+        self.term(out);
+
+        while let Some(op) = self.next_char() {
+            if op == '+' || op == '-' {
+                self.term(out);
+                out.push(op);
+            } else {
+                panic!("Unexpected symbol '{}'", op);
+            }
+        }
+    }
+
+    fn term(&mut self, out: &mut String) {
+        match self.next_char() {
+            Some(ch) if ch.is_digit(10) => out.push(ch),
+            Some(ch) => panic!("Unexpected symbol '{}'", ch),
+            None => panic!("Unexpected end of string"),
+        }
+    }
+}
+
+pub fn main() {
+    let infix = String::from("2+3");
+    let mut intr = Interpreter::new(&infix);
+    let mut postfix = String::new();
+    intr.interpret(&mut postfix);
+    assert_eq!(postfix, "23+");
+
+    let infix = String::from("1-2+3-4");
+    intr = Interpreter::new(&infix);
+    postfix.clear();
+    intr.interpret(&mut postfix);
+    assert_eq!(postfix, "12-3+4-");
+}
+```
+
+## Discussion
+
+There may be a wrong perception that the Interpreter design pattern
+is about design grammars for formal languages and implementation
+of parsers for these grammars. In fact, this pattern is about expressing
+problem instances in a more specific way and implementing
+functions/classes/structs solving these problem instances.
+Rust language has `macro_rules!` that allows to define special
+syntax and rules how to expand this syntax into source code.
```

Comment:
I think that "how to" is not correct here.

Reply:
@MarcoIeni Why "how to" is not correct? This is from Rust by Examples book: 
> instead of generating a function call, macros are expanded into source code that gets compiled with the rest of the program

But how, I mean according to what rules, it will be expanded depends on us, e.g., on how we define the rules.

Reply:
no, sorry, I mean from an English grammar point of view.
Maybe it is better:

```suggestion
syntax and rules on how to expand this syntax into source code.
```